### DISCUSSION TITLE
Auto-generated schema files don't work out-of-the-box

### DISCUSSION THREAD

**Discussion by iMacTia** [flipp-oss/deimos#150] [Q&A] [answered]
We might be doing something wrong here, but basically we have our schemas under a certain namespace (e.g. `org.company.my_schemas`).
This means that schema files need to go into a nested folder structure (e.g. `org/company/my_schemas`).

The auto-generated Ruby class gets placed alongside the schema file, but the only module in the file is module Schemas.
This breaks Rails auto-loading, which won't be able to find these classes. So we manually move them to a different location, but that means every time the `rake deimos:generate_schema_classes` is executed, it will re-create all the schemas that were previously moved.

The solution here would be to either include the schema namespace in the ruby file modules, or create the ruby file in the root (specified in the Deimos config).

**Comment by dorner** [comment]
I actually haven't used this feature much myself yet as it's relatively new and was contributed by a coworker. @DeeChau can you weigh in on this question?

**Comment by DeeChau** [comment, marked as answer]
Sure thing @dorner,

Hey @iMacTia, generated schema classes are generated under `Deimos.config.schema.generated_class_path` which defaults to the path `app/lib/schema_classes`, so the nested folder structure `org/company/my_schemas` are be created with that in mind.

> The auto-generated Ruby class gets placed alongside the schema file, but the only module in the file is module Schemas.

Can you provide me with the `config.schema.generated_class_path` and `config.schema.path` to help me understand where these are located?

> The solution here would be to either include the schema namespace in the ruby file modules, or create the ruby file in the root (specified in the Deimos config).

I considered including the schema namespace in the ruby file modifiers during development and it is possible. As for creating the ruby file in the root, do you mean at the `config.schema.generated_class_path`?

**Reply by iMacTia** [reply]
> Can you provide me with the `config.schema.generated_class_path` and `config.schema.path` to help me understand where these are located?

Sure! This is my `schema` config block:

```ruby
schema do
  path 'app/event_streams/schemas'
  generated_class_path 'app/event_streams/schemas'
end
```

So supposing I have an `org.company.my_schema` schema, that would go into `app/event_streams/schemas/org/company/my_schema.avsc`, so far so good 👍
Now when I run `rake deimos:generate_schema_classes`, a new `app/event_streams/schemas/org/company/my_schema.rb` file is auto-generated, but the content of the file is:

```ruby
module Schemas
  class MySchema < Deimos::SchemaClass::Enum
   ...
  end
end
```

So because the modules namespace in the file doesn't match the location of the file, Rails is unable to autoload it, and I have to manually move it to `app/event_streams/schemas/my_schema.rb`.

> I considered including the schema namespace in the ruby file modifiers during development and it is possible. As for creating the ruby file in the root, do you mean at the `config.schema.generated_class_path`?

Yes, that would be it, assuming that path points to a `something/schemas` folder, the namespacing of the file would match and Rails would be able to autoload the schema correctly. It's worth noting though that creating the Ruby file in the root may cause naming clashes (e.g. same schema name under different namespaces).

**Reply by DeeChau** [reply]
@iMacTia 
I can see that you're intentionally using the same directory for both the schemas and the schema classes 👍 

When I try using the same `config.schema.path` and `config.schema.generated_class_path`, i unfortunately don't see any problems with autoloading in my rails app 🤔 

I'm afraid that I don't understand why this is not loading it properly for your app. The code in question that was added to load these classes in at startup is [here](https://github.com/flipp-oss/deimos/blob/5f2fbd4ce998244ed45b1c49443067d31d822b5f/lib/deimos/config/configuration.rb#L34-L43) so the `module Schemas` should not have to care about the directory structure that the files reside in

The fact that Rails is complaining about this is... confusing. Can you post the Rails Autoload error-message that you are seeing?
Ruby + Rails versions would be welcome too to help figure out these errors.

Otherwise, generating everything to `config.schema.generated_class_path` might be a worthwhile endeavour to help solve the immediate problem, seeing as clashing schema names doesn't seem like great practice with Avro Schema naming.

**Reply by iMacTia** [reply]
I was getting `uninitialized constant Schemas::MySchema`, but after you pointed me to the code that was loading the schemas folder, I noticed that it only runs when `Deimos.config.schema.use_schema_classes` is set to `true`.
I checked my config and I was only setting `use_schema_classes` in producers, which explains why the schemas were not loaded!

That worked 🙏! But I'm still not sure about what the difference between the schema-level and producer-level `use_schema_classes ` setting is?

**Reply by dorner** [reply]
@iMacTia The schema-level setting is the general one for the application. The producer and consumer level are overrides. So you can set it to true for the application but false for specific producers, or vice versa.

**Reply by iMacTia** [reply]
aha! Thanks @dorner, so can I safely remove the `use_schema_classes` config from producers?

**Reply by dorner** [reply]
If you have it set at the app-level, yes!

**Reply by DeeChau** [reply]
Yeap @iMacTia! We probably need to be more clear in our documentation

**Reply by iMacTia** [reply]
Thank you both for helping out 🙌!
